### PR TITLE
Integrated Classic support

### PR DESCRIPTION
**Pull request by inyawfaze**
Native support for Classic this time.

HelpTips and several widgets will not get loaded for Classic, via exclusion from it's .toc.
Unsupported frames (focus/boss) are excluded for Classic.

Exclusion of these features from the baseline const, defaults and builder tables & conditionally adding them in for the Retail version.

Some util functions are added in esp. regarding auras. 
I spent quite some time researching how to retrieve the buff index needed for the aura tooltips in Classic, an on demand util function seemed reasonable, as the alternative requires constantly mapping the index on every buff change.

Following this approach now only available features will be visible in the addon menu.
Role icons are disabled for Classic for now, as their LibGroupInfo requirement from Cell is not loaded for Classic. This should be able to get reimplemented fairly easily.

**Typings**: definitely not my strongsuit. Had to change some enums to classes as they have keys included conditionally. At least that's what my LuaLS suggested, I hope thats the correct approach. These changes will definitely add some warnings. 

**Retail**: every feature I took out for Classic worked for me in Retail. I don't think there should be any regression/breaking changes in this.

I think this approach is way healthier than my last one, I'm happy I redid this (and will keep working on it if it's not up to standard)!

TLDR/Commit Message:
 - Seperate .toc which does not load HelpTips and some widgets
 - adjusted constants and defaults
 - removes unsupported widgets from builder
 - removes unsupported health formats
 - helper functions for missing AuraUtil
 - seperate aura tooltip handling
 - several util functions

### CODE REVIEW

**Comment by inyawfaze** [comment]
Yeah, the lua ci runner does not seem to approve my changes. I'm not too sure if I can fix those myself, but I'll take a look and learn about it. 😅

**Comment by Krealle** [comment]
I will try to take a look over this later today - but changing enums to classes is a big no-no.

Enums essentially describe an input to be x value out of y enum table.
Whilst a class is used to define a table.

So if you have a function that expects a string to be one of the entries in an enum, then you can type it as such, but if it becomes a class, then it would expect it to be a table.

File: Data/Defaults.lua [hunk 1]
```diff
@@ -756,6 +798,17 @@ Defaults.Widgets = {
...
+if CUF.vars.isRetail then
+    ---@type ShieldBarWidgetTable
+    Defaults.Widgets.shieldBar = {
+        enabled = false,
+        frameLevel = 9,
+        point = "RIGHT",
+        reverseFill = false,
+        overShield = false,
+    }
+end
```
Comment (Krealle) [inline Data/Defaults.lua:810]: Shouldn't Class bar and heal absorb also be gated here?

File: Data/Defaults.lua [hunk 2]
```diff
@@ -995,7 +1007,53 @@ Defaults.Layouts = {
         healthBarColorType = CUF.constants.UnitButtonColorType.CELL,
         healthLossColorType = CUF.constants.UnitButtonColorType.CUSTOM,
     },
-    boss = {
+}
+
+-- Focus was introduced with TBC
+if not CUF.vars.isVanilla then
+    Defaults.Layouts.focus = {
```
Comment (Krealle) [inline Data/Defaults.lua:1014]: You need to use the const to make the typing happy, it's a bit dumb xd
```suggestion
    Defaults.Layouts[CUF.constants.UNIT.FOCUS]
```

File: Data/Constants.lua [hunk 3]
```diff
@@ -325,10 +347,18 @@ const.BlizzardFrameTypes = {
...
+
+-- Focus was introduced with TBC
+if not CUF.vars.isVanilla then
+    table.insert(const.BlizzardFrameTypes, CUF.constants.UNIT.FOCUS)
+end
+
+-- Boss frames were intoduced with Wrath
+if not CUF.vars.isVanilla and not CUF.vars.isBCC then
+    table.insert(const.BlizzardFrameTypes, CUF.constants.UNIT.BOSS)
+end
```
Comment (Krealle) [inline Data/Constants.lua:364]: NIT: inserting like this will mess up the order in which they are displayed.

**Review by Krealle — CHANGES_REQUESTED** [review]
Okay, lots of stuff going on here. Overall the direction you went with this PR is in the right one.
There are things that need to addressed and changed however.

As for typing I've tried to point out the areas where it went wrong and how to resolve them.

As for the HelpTips - The file should still be loaded, but just make dummy functions for Acknowledge and Show funcs so we can reduce complexity. Could also just be a HelpTipVanilla.lua file.

I have some reservations about how the Defaults are handled, but that's not really something that I'm *that* concerned with atm.

If you have any questions about any of the comments feel free to ask them.

**Comment by Krealle** [comment]
I also think you should add `---@diagnostic disable: inject-field` to the top of `Constants.lua`. Should help with removing all the typing nightmares there.

**Comment by SekSion** [comment]
Is there an update for this?

**Comment by Krealle** [comment]
Closing due to inactivity and changes being made here #281
